### ISSUE TITLE
Should the user be allowed to define and explicitly call a module-level `init` function?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#16760]
Currently we disallow explicit calls to module `deinit` functions. Should we do the same for module `init` functions?

Furthermore, should module `init` functions be allowed at all, as we currently do not invoke them implicitly?

Module init functions are discussed in #5792
Remotely related issues: &nbsp; #6726 &nbsp; #11445 &nbsp; https://github.com/chapel-lang/chapel/issues/13897#issuecomment-529994657

For example, this program compiles and prints "init":
```chpl
module M {
  proc init() {
    writeln("init");
  }
}
module Main {
  use M;
  proc main {
    init();
  }
}
```

If I comment out the call to init() in main(), init() will not be invoked.

We have at least these tests doing this as of this writing. I plan to rename their init functions soon:
```
test/modules/diten/moduleNotInitialized.chpl
test/modules/diten/moduleNotInitialized2.chpl
```

**Comment by bradcray** [comment]
Interesting questions.  My intuition is that we should probably either support them (in the sense of "automatically call them when a module is initialized), outlaw, or warn about them.  I lean a bit towards supporting them, though there's a bit of a question in my mind about when to call them (perhaps "after all module-scope declarations and code have run?").

**Comment by vasslitvinov** [comment]
The order of execution is indeed a tricky quesiton. On the one hand, some might expect execution in reading order. However, `init` is the opposite of `deinit` and the latter is called the first thing, out of order w.r.t. deinitializing the variables:
```chpl
var myRecord1 ...;     // deinitialized 3rd
proc deinit() {...}    // called 1st
var myRecord2 ...;     // deinitialized 2nd
```

To mirror that, we'd need to call init() out of order as well:
```chpl
writeln("initialize A");                    // printed 1st
proc init() { writeln("in module init"); }  // printed 3rd
writeln("initialize B");                    // printed 2nd
```

Having written it out, out-of-order-ness makes sense to me.

**Comment by bradcray** [comment]
> out-of-order-ness makes sense to me.

I agree.  And I think calling init() last rather than where it was defined makes sense in the same way that main() is called at a certain point rather than where it was defined.  It's not executable code, it's a routine to be called, so the call can occur whenever we want it to be.
